alibi_detect/utils/_types.py

```diff
@@ -51,3 +53,33 @@ def _validate(cls: Type, val: Any, field: ModelField) -> np.ndarray:
         return np.asarray(val, dtype=dtype_field.type_)
     else:
         return np.asarray(val)
+
+
+# Optional dep dependent type definitions. Two sets of items are defined here:
+# 1. Tuples of actual objects, conditional on installed ops deps, for use in isinstance()'s, including in pydantic.
+# 2. Type unions for type hints. Note: Not currently used properly by mypy since there are no stubs for tf, sklearn etc
+
+# 1. Tuples
+supported_models_tf, supported_models_torch, supported_models_sklearn = (), (), ()  # type: ignore
+supported_optimizers_tf, supported_optimizers_torch = (), ()  # type: ignore
+if has_tensorflow:
+    import tensorflow as tf
+    supported_models_tf = (tf.keras.Model, )  # type: ignore
+    supported_optimizers_tf = (tf.keras.optimizers.Optimizer, )  # type: ignore
+if has_pytorch:
+    import torch
+    supported_models_torch = (torch.nn.Module, torch.nn.Sequential)  # type: ignore
+    supported_optimizers_torch = (torch.optim.Optimizer, )  # type: ignore
+supported_models_sklearn = (BaseEstimator, )  # type: ignore
+supported_models_all = supported_models_tf + supported_models_torch + supported_models_sklearn
+supported_optimizers_all = supported_optimizers_tf + supported_optimizers_torch
+
+# 2. Type unions
+model_types_tf = Type['tf.keras.Model']
+model_types_torch = Union['torch.nn.Module', 'torch.nn.Sequential']
+model_types_sklearn = Type[BaseEstimator]  # no ForwardRef since sklearn a core dep
+optimizer_types_tf = Type['tf.keras.optimizers.Optimizer']
+optimizer_types_torch = Union['torch.optim.Optimizer']
```

Comment:
`Union` -> `Type`?

Reply:
Btw what's the difference between `Type[SomeClass]` and just `SomeClass` when used in type hints?

Reply:
Without `Type` I get the following mypy error when `model_types_tf` is added to `model_types_all`:

```
error: Variable "alibi_detect.utils._types.model_types_tf" is not valid as a type  [valid-type]
```

To be honest I'm wondering if we should just remove lines 77 onwards for now, since we are not using them elsewhere in the code yet anyway. I've also just realised that `models_types_all` might no actually work properly anyway, since we are still defining this at runtime hence mixing dynamic and static typing again.

Reply:
Removed (for now) in e4ed219fb6b8d559eac503bbcd0200d78301e811.